src/server/gc_worker/gc_worker.rs

```diff
@@ -247,40 +252,128 @@ struct KeysInRegions<R: Iterator<Item = Region>> {
         loop {
             let region = self.regions.peek()?;
             let key = self.keys.peek()?.as_encoded().as_slice();
             if key < region.get_start_key() {
                 self.keys.next();
             } else if region.get_end_key().is_empty() || key < region.get_end_key() {
-                return self.keys.next();
+                return self.keys.next().map(|key| (key, region.clone()));
             } else {
                 self.regions.next();
             }
         }
     }
 }
 
+// If `region_or_provider` is Region, it means all keys are located in the same
+// region. If `region_or_provider` is RegionInfoProvider, then we should find
+// all regions of thoes keys.
+// We return an iterator which yeild item of `Key` and the region taht the key
+// is located. We also return the reion of the first key.
 fn get_keys_in_regions(
+    store_id: u64,
     keys: Vec<Key>,
-    regions_provider: Option<(u64, Arc<dyn RegionInfoProvider>)>,
-) -> Result<Box<dyn Iterator<Item = Key>>> {
-    if keys.len() >= 2 {
-        if let Some((store_id, region_info_provider)) = regions_provider {
-            let start = keys.first().unwrap().as_encoded();
-            let end = keys.last().unwrap().as_encoded();
-            let regions = box_try!(region_info_provider.get_regions_in_range(start, end))
-                .into_iter()
-                .filter(move |r| find_peer(r, store_id).is_some())
-                .peekable();
+    region_or_provider: Either<Region, Arc<dyn RegionInfoProvider>>,
+) -> Result<(Box<dyn Iterator<Item = (Key, Region)>>, Region)> {
+    assert!(!keys.is_empty());
+
+    match region_or_provider {
+        Either::Left(region) => {
+            let region_clone = region.clone();
+            Ok((
+                Box::new(keys.into_iter().map(move |k| (k, region.clone()))),
+                region_clone,
+            ))
+        }
+        Either::Right(region_provider) => {
+            if keys.len() >= 2 {
+                let start = keys.first().unwrap().as_encoded();
+                let end = keys.last().unwrap().as_encoded();
+                let mut regions = box_try!(region_provider.get_regions_in_range(start, end))
+                    .into_iter()
+                    .filter(move |r| find_peer(r, store_id).is_some())
+                    .peekable();
+
+                let region = match regions.peek() {
+                    Some(region) => region.clone(),
+                    None => {
+                        return Err(box_err!(
+                            "No region has been found for range [{:?}, {:?})",
+                            start,
+                            end
+                        ));
+                    }
+                };
 
-            let keys = keys.into_iter().peekable();
-            return Ok(Box::new(KeysInRegions { keys, regions }));
+                let keys = keys.into_iter().peekable();
+                Ok((Box::new(KeysInRegions { keys, regions }), region))
```

Comment:
Did you consider the case that the first key isn't in any region currently in this TiKV instance, in which case we expect to skip the key?

Reply:
I have made some modifications for this. PTAL.